core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java

```diff
@@ -65,4 +82,65 @@ public int totalOperations() {
     public void close() throws IOException {
         onClose.close();
     }
+
+    /**
+     * A wrapper of {@link FixedBitSet} but allows to check if all bits are set in O(1).
+     */
+    private static final class CountedBitSet {
+        private short onBits;
+        private final FixedBitSet bitset;
+
+        CountedBitSet(short numBits) {
+            assert numBits > 0;
+            this.onBits = 0;
+            this.bitset = new FixedBitSet(numBits);
+        }
+
+        boolean getAndSet(int index) {
+            assert index >= 0;
+            boolean wasOn = bitset.getAndSet(index);
+            if (wasOn == false) {
+                onBits++;
+            }
+            return wasOn;
+        }
+
+        boolean hasAllBitsOn() {
+            return onBits == bitset.length();
+        }
+    }
+
+    /**
+     * Sequence numbers from translog are likely to form contiguous ranges, thus using two tiers can reduce memory usage.
+     */
+    static final class SeqNumSet {
```

Comment:
Could we call this `SequenceNumberSet` instead? Also, this idea feels more widely applicable. Can it be used elsewhere?

Reply:
This should be used by `MultiSnapshot` only.